haystack/schema.py

```diff
@@ -758,7 +776,7 @@ def __str__(self):
         return f"<MultiLabel: {self.to_dict()}>"
 
 
-def _pydantic_dataclass_from_dict(dict: dict, pydantic_dataclass_type) -> Any:
+def _pydantic_dataclass_from_dict(dict: "dict", pydantic_dataclass_type) -> Any:
```

Comment:
Why are we doing `dict: "dict"` instead of `data: Dict`? 
More often than not it's not a great idea to overwrite builtins, we must avoid it.

Same thing for other occurances of this.

Reply:
Yup I'm with you on not overwriting builtins. I'd be happy to rename the variable to something else, but I'm afraid this might cause breaking changes. I'd be happy to switch to `Dict` in this case.

Reply:
Ah right, that might break stuff. 🤦 

I think we can still leave the variable as `dict` but set its type as `Dict`. I'd expect that won't cause issues.